netty/src/main/java/io/grpc/netty/GrpcHttp2Headers.java

```diff
@@ -0,0 +1,146 @@
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+package io.grpc.netty;
+
+import io.netty.handler.codec.http2.Http2Headers;
+import io.netty.util.AsciiString;
+
+import java.util.Iterator;
+import java.util.Map.Entry;
+import java.util.NoSuchElementException;
+
+/**
+ * A custom implementation of Http2Headers that only includes methods used by gRPC.
+ */
+final class GrpcHttp2Headers extends AbstractHttp2Headers {
+
+  private final AsciiString[] normalHeaders;
+  private final AsciiString[] preHeaders;
+  private static final AsciiString[] EMPTY = new AsciiString[]{};
+
+  static GrpcHttp2Headers clientHeaders(byte[][] serializedMetadata, AsciiString authority,
+      AsciiString path, AsciiString method, AsciiString scheme, AsciiString userAgent) {
+    AsciiString[] preHeaders = new AsciiString[] {
+        Http2Headers.PseudoHeaderName.AUTHORITY.value(), authority,
+        Http2Headers.PseudoHeaderName.PATH.value(), path,
+        Http2Headers.PseudoHeaderName.METHOD.value(), method,
+        Http2Headers.PseudoHeaderName.SCHEME.value(), scheme,
+        Utils.CONTENT_TYPE_HEADER, Utils.CONTENT_TYPE_GRPC,
+        Utils.TE_HEADER, Utils.TE_TRAILERS,
+        Utils.USER_AGENT, userAgent,
+    };
+    return new GrpcHttp2Headers(preHeaders, serializedMetadata);
+  }
+
+  static GrpcHttp2Headers serverHeaders(byte[][] serializedMetadata) {
+    AsciiString[] preoHeaders = new AsciiString[] {
+        Http2Headers.PseudoHeaderName.STATUS.value(), Utils.STATUS_OK,
+        Utils.CONTENT_TYPE_HEADER, Utils.CONTENT_TYPE_GRPC,
+    };
+    return new GrpcHttp2Headers(preoHeaders, serializedMetadata);
+  }
+
+  static GrpcHttp2Headers serverTrailers(byte[][] serializedMetadata) {
+    return new GrpcHttp2Headers(EMPTY, serializedMetadata);
+  }
+
+  private GrpcHttp2Headers(AsciiString[] preHeaders, byte[][] serializedMetadata) {
+    normalHeaders = new AsciiString[serializedMetadata.length];
+    for (int i = 0; i < normalHeaders.length; i++) {
+      normalHeaders[i] = new AsciiString(serializedMetadata[i], false);
+    }
+    this.preHeaders = preHeaders;
+  }
+
+  @Override
+  public Iterator<Entry<CharSequence, CharSequence>> iterator() {
+    return new Itr();
+  }
+
+  @Override
+  public int size() {
+    return (normalHeaders.length + preHeaders.length) / 2;
+  }
+
+  private class Itr implements Iterator<Entry<CharSequence, CharSequence>> {
+    private int normalIdx;
+    private int preIdx;
+    private AsciiString[] currentArray = preHeaders;
+
+    @Override
+    public boolean hasNext() {
+      return preIdx < preHeaders.length || normalIdx < normalHeaders.length;
```

Comment:
I wonder if we can reduce the number of conditions we're checking throughout by using 2 extra objects. This also includes buchgr@'s idea for implementing the `Entry` interface:

```java
private class Itr implements Iterator<Entry<CharSequence, CharSequence>> {
  private GroupIter group;
  Itr() {
    group = new GroupIter(preHeaders, normalHeaders.length > 0 ? new GroupIter(normalHeaders, null) : null);
  }

  @Override
  public boolean hasNext() {
    return group.hasNext();
  }

  @Override
  public Entry<CharSequence, CharSequence> next() {
    return group.next();
  }

  @Override
  public void remove() {
    throw new UnsupportedOperationException();
  }

  private final class GroupIter implements Entry<CharSequence, CharSequence> {
    private final GroupIter next;
    private final AsciiString[] headers;
    private int idx = 0;

    GroupIter(AsciiString[] headers, @Nullable GroupIter next) {
      this.headers = headers;
      this.next = next;
    }

    @Override
    public CharSequence setValue(CharSequence value) {
      throw new UnsupportedOperationException();
    }

    @Override
    public CharSequence getValue() {
      return headers[idx];
    }

    @Override
    public CharSequence getKey() {
      return headers[idx - 1];
    }

    boolean hasNext() {
      if (idx < headers.length) {
        return true;
      }
      return next != null;
    }

    Entry<CharSequence, CharSequence> next() {
      if (idx >= headers.length) {
        if (next == null) {
	  throw new NoSuchElementException();
        }
        group = next;
        return group.next();
      }

      idx += 2;
      return this;
    }
  }
}
```